modules/cudacodec/src/video_writer.cpp

```diff
@@ -68,12 +67,48 @@ bool operator==(const EncoderParams& lhs, const EncoderParams& rhs)
             rhs.averageBitRate, rhs.maxBitRate, rhs.targetQuality, rhs.gopLength);
 };
 
+class FFmpegVideoWriter : public EncoderCallback
+{
+public:
+    FFmpegVideoWriter(const String& fileName, const Codec codec, const int fps, const Size sz, const int idrPeriod);
+    ~FFmpegVideoWriter();
+    void onEncoded(const std::vector<std::vector<uint8_t>>& vPacket);
+    void onEncodingFinished();
+private:
+    cv::VideoWriter writer;
+};
+
+FFmpegVideoWriter::FFmpegVideoWriter(const String& fileName, const Codec codec, const int fps, const Size sz, const int idrPeriod) {
+    if (!videoio_registry::hasBackend(CAP_FFMPEG))
+        CV_Error(Error::StsNotImplemented, "FFmpeg backend not found");
+    const int fourcc = codec == Codec::H264 ? cv::VideoWriter::fourcc('a', 'v', 'c', '1') : cv::VideoWriter::fourcc('h', 'e', 'v', '1');
```

Comment:
What about VP8, VP9, MPEG4, others?

Reply:
The Nvidia Video Codec SDK only [supports ](https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new) H264, H265 and AV1 encoding, with `cudacodec::VideoWriter` only supporting H264[5].  `cudacodec::VideoWriter` was introduced/fixed/updated before Video Codec SDK v12.0 where AV1 encoding was introduced.  Because AV1 is only supported on Ada GPU's and I don't have access to one for testing I have not yet added that functionality.

Reply:
Sorry, I screwed up with decoder. Please add assertion for codec then.

Reply:
This is taken care of in `CodecGuid` which is called during the initialization of the `VideoWriterImpl`.